### PR TITLE
[#10][#29] [BACKEND] [FRONTEND] As a user, I can query the search results across all reports

### LINKED ISSUE #10: [BACKEND] As a user, I can query the search results across all reports
###  Why

User should be able to query all reports at once.

* Search by keyword
* Filter by min_ad_count, max_ad_count.

### Who Benefits

User

### PR DESCRIPTION
**Pull request by ankitkalia1195**
#10 #29 

## What happened

✅ Added FRONTEND and BACKEND for searching and filtering the keywords 

## Insight

* In this ticket, only searching from keywords name, and not the content in json field. (keyword search results)
* For searching in json field, I will add it later.
* Also added a filter for links count.(min and max)

1. Created a new class `KeywordQuery`, for searching and filtering. 
```
   defmodule KeywordQuery
     def all(user, params)
     end
  end
```

2. Created KeywordSearchForm class with embedded_struct, for rendering the form

```
    defmodule KeywordSearchForm
       embedded_struct do
       end
    end

```
 
## Proof Of Work

https://user-images.githubusercontent.com/11224179/122343522-fc87a680-cf6f-11eb-9ae7-d0c24d9aa5df.mov

### CODE REVIEW

File: lib/google_crawler/search/keyword_query.ex [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  alias GoogleCrawler.Repo
+  alias GoogleCrawler.Search.{Keyword, KeywordSearchForm}
+
+  def all(user, params) do
+    Keyword
+    |> where([k], k.user_id == ^user.id)
+    |> filter_by_term(params)
+    |> filter_by_links_count(params)
+    |> order_by(desc: :inserted_at)
+    |> Repo.all()
```
Comment (rosle) [inline lib/google_crawler/search/keyword_query.ex]: Should we move Repo command out and let the query only have the queries? We can do the Repo.all in context instead 🤔

File: lib/google_crawler_web/templates/keyword/index.html.eex [hunk 2]
```diff
@@ -1,34 +1,59 @@
...
+            </tr>
+          </thead>
+          <tbody>
+        <%= for keyword <- @keywords do %>
+          <tr>
+            <td><%= link keyword.name, to: Routes.keyword_path(@conn, :show, keyword.id) %></td>
+            <td><%= keyword.status %></td>
+            <%= if keyword.result do %>
+              <td class="count"><%= keyword.result.links_count %></td>
+              <td class="count"><%= keyword.result.top_ads |> Enum.count() %></td>
```
Comment (hanam1ni) [inline lib/google_crawler_web/templates/keyword/index.html.eex]: Let's use the function directly if it only single statement
```suggestion
              <td class="count"><%=  Enum.count(keyword.result.top_ads) %></td>
```

File: lib/google_crawler_web/templates/keyword/index.html.eex [hunk 3]
```diff
@@ -1,34 +1,59 @@
...
-    <tr>
-      <td><%= link keyword.name, to: Routes.keyword_path(@conn, :show, keyword.id) %></td>
-      <td><%= keyword.status %></td>
-      <%= if keyword.result do %>
-        <td class="count"><%= keyword.result.links_count %></td>
-        <td class="count"><%= keyword.result.top_ads |> Enum.count() %></td>
-        <td class="count"><%= keyword.result.non_ads |> Enum.count() %></td>
+        <div class="form-group">
+          <%= label :min_links_count, "Links Count(Min)" %>
+          <%= select f, :min_links_count, links_count_select_options() %>
```
Comment (hanam1ni) [inline lib/google_crawler_web/templates/keyword/index.html.eex]: How about updating this to `number_input`?

File: lib/google_crawler/search/keyword_search_form.ex [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  embedded_schema do
+    field :term, :string
+    field :min_links_count, :integer
+    field :max_links_count, :integer
+  end
+
+  def changeset(keyword_search_form, attrs) do
```
Comment (hanam1ni) [inline lib/google_crawler/search/keyword_search_form.ex]: How about set default params for form?
```suggestion
  def changeset(keyword_search_form \\ %__MODULE__{}, attrs) do
```

So we can use it like this
```elixir
KeywordSearchForm.changeset(params["filter"] || %{})
```

File: lib/google_crawler_web/controllers/keyword_controller.ex [hunk 5]
```diff
@@ -1,12 +1,17 @@
...
+  alias GoogleCrawler.Search.{Keyword, KeywordQuery, KeywordSearchForm}
 
-  def index(conn, _params) do
-    keywords = Search.list_keywords(conn.assigns[:current_user])
-    render(conn, "index.html", keywords: keywords)
+  def index(conn, params) do
+    search_form_changeset =
+      KeywordSearchForm.changeset(%KeywordSearchForm{}, params["filter"] || %{})
+
+    search_form_params = Ecto.Changeset.apply_changes(search_form_changeset)
```
Comment (hanam1ni) [inline lib/google_crawler_web/controllers/keyword_controller.ex]: How about moving this logic to `GoogleCrawler.Search.KeywordSearchForm` to make controller cleaner?
```elixir
# lib/google_crawler/search/keyword_search_form.ex
def new(params) do
  %__MODULE__{}
  |> changeset(params)
  |> Ecto.Changeset.apply_changes()
end

defp changeset() do
  # ...
end

# lib/google_crawler_web/controllers/keyword_controller.ex
def index() do
  search_form = KeywordSearchForm.new(params)
end
```

File: lib/google_crawler/search/keyword_query.ex (hunk 1 above)
Comment (ankitkalia1195) [reply lib/google_crawler/search/keyword_query.ex]: Fixed

File: lib/google_crawler_web/templates/keyword/index.html.eex (hunk 2 above)
Comment (ankitkalia1195) [reply lib/google_crawler_web/templates/keyword/index.html.eex]: Fixed

File: lib/google_crawler_web/templates/keyword/index.html.eex (hunk 3 above)
Comment (ankitkalia1195) [reply lib/google_crawler_web/templates/keyword/index.html.eex]: Done

File: lib/google_crawler_web/controllers/keyword_controller.ex (hunk 5 above)
Comment (ankitkalia1195) [reply lib/google_crawler_web/controllers/keyword_controller.ex]: Nice one, thanks for the suggestion. I extracted it out of controller

**Comment by ankitkalia1195** [comment]
@hanam1ni @rosle Done with all the changes

File: lib/google_crawler_web/controllers/keyword_controller.ex [hunk 6]
```diff
@@ -1,12 +1,16 @@
...
+  alias GoogleCrawler.Repo
+  alias GoogleCrawler.Search
+  alias GoogleCrawler.Search.{Keyword, KeywordQueryParams}
 
-  def index(conn, _params) do
-    keywords = Search.list_keywords(conn.assigns[:current_user])
-    render(conn, "index.html", keywords: keywords)
+  def index(conn, params) do
+    %{query_params: query_params, changeset: changeset} =
+      KeywordQueryParams.new(params["filter"] || %{})
```
Comment (rosle) [inline lib/google_crawler_web/controllers/keyword_controller.ex:10]: I'm quite unsure about this 🤔 

Maybe you can take a look at the params validator on Toyota wallet? - https://github.com/nimblehq/gcp-api/blob/master/apps/gcpf_common/lib/gcpf_common/params_validator.ex. 

But you want to use both results - params and changeset here, right? 🤔  Then Toyota wallet's one might not fit here.

**Review by rosle — APPROVED** [review]

File: lib/google_crawler_web/controllers/keyword_controller.ex (hunk 6 above)
Comment (ankitkalia1195) [reply lib/google_crawler_web/controllers/keyword_controller.ex:10]: @rosle Yes I need both the results here, one for easily rendering the form(changeset), as form_for helper supports only changeset, and another for filtering and passing to KeywordQuery.

Comment (rosle) [reply lib/google_crawler_web/controllers/keyword_controller.ex:10]: Aha, I see 👀  Actually for form_for, we can also use with atom like this `form_for :filter, ...` but we will need to populate the input value by ourself by, maybe `value: get_in(params, ["filter", "term"])`

Comment (ankitkalia1195) [reply lib/google_crawler_web/controllers/keyword_controller.ex:10]: @rosle Yes I want to avoid that, that is the reason I went this way. :)

**Review by hanam1ni — APPROVED** [review]

**Pull request merged**